src/main/java/org/traccar/api/security/OpenIDProvider.java

```diff
@@ -0,0 +1,245 @@
+import com.nimbusds.oauth2.sdk.auth.Secret;
+import com.nimbusds.oauth2.sdk.auth.ClientSecretBasic;
+import com.nimbusds.oauth2.sdk.auth.ClientAuthentication;
+import com.nimbusds.oauth2.sdk.token.BearerAccessToken;
+import com.nimbusds.oauth2.sdk.id.State;
+import com.nimbusds.oauth2.sdk.id.ClientID;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponse;
+import com.nimbusds.openid.connect.sdk.Nonce;
+import com.nimbusds.openid.connect.sdk.OIDCTokenResponseParser;
+import com.nimbusds.openid.connect.sdk.UserInfoResponse;
+import com.nimbusds.openid.connect.sdk.UserInfoRequest;
+import com.nimbusds.openid.connect.sdk.AuthenticationRequest;
+
+import com.nimbusds.openid.connect.sdk.claims.UserInfo;
+
+public class OpenIDProvider {
+    public final Boolean force;
+    private final ClientID clientId;
+    private final Secret clientSecret;
+    private URI callbackUrl;
+    private URI authUrl;
+    private URI tokenUrl;
+    private URI userInfoUrl;
+    private URI baseUrl;
+    private final String adminGroup;
+
+    private Config config;
+    private LoginService loginService;
+    private Storage storage;
+
+    @Inject
+    public OpenIDProvider(Config config, LoginService loginService, Storage storage) {
+        force = config.getBoolean(Keys.OIDC_FORCE);
+        clientId = new ClientID(config.getString(Keys.OIDC_CLIENTID));
+        clientSecret = new Secret(config.getString(Keys.OIDC_CLIENTSECRET));
+
+        this.config = config;
+        this.storage = storage;
+        this.loginService = loginService;
+
+        try {
+            callbackUrl = new URI(config.getString(Keys.WEB_URL, "") + "/api/session/openid/callback");
+            authUrl = new URI(config.getString(Keys.OIDC_AUTHURL, ""));
+            tokenUrl = new URI(config.getString(Keys.OIDC_TOKENURL, ""));
+            userInfoUrl = new URI(config.getString(Keys.OIDC_USERINFOURL, ""));
+            baseUrl = new URI(config.getString(Keys.WEB_URL, ""));
+        } catch (URISyntaxException e) {
+        }
+
+        adminGroup = config.getString(Keys.OIDC_ADMINGROUP);
+    }
+
+    public URI createAuthRequest() {
+        AuthenticationRequest request = new AuthenticationRequest.Builder(
+                new ResponseType("code"),
+                new Scope("openid", "profile", "email", "groups"),
+                this.clientId,
+                this.callbackUrl)
+                .endpointURI(this.authUrl)
+                .state(new State())
+                .nonce(new Nonce())
+                .build();
+
+        return request.toURI();
+    }
+
+    private OIDCTokenResponse getToken(AuthorizationCode code) {
+        // Credentials to authenticate us to the token endpoint
+        ClientAuthentication clientAuth = new ClientSecretBasic(this.clientId, this.clientSecret);
+        AuthorizationGrant codeGrant = new AuthorizationCodeGrant(code, this.callbackUrl);
+
+        TokenRequest request = new TokenRequest(this.tokenUrl, clientAuth, codeGrant);
```

Comment:
Can this be reused? That way we don't need to keep so many instance variables.

Reply:
clientAuth can be and thats the only place we use clientSecret, so I've moved that to a class property.